src/jit.rs

```diff
@@ -755,78 +801,82 @@ impl<'a> JitMemory<'a> {
-                    // // For JIT, syscalls in use MUST be registered at compile time. They can be
-                    // // updated later, but not created after compiling (we need the address of the
-                    // // syscall function in the JIT-compiled program).
-                    // if let Some(syscall) = syscalls.get(&(insn.imm as u32)) {
-                    //     if syscall.verifier.is_some() {
-                    //         return Err(Error::new(ErrorKind::Other,
-                    //                        format!("[JIT] Error: syscall verifier function not supported by jit (id: {:#x})",
-                    //                                insn.imm as u32)));
-                    //     }
-                    //     // We reserve RCX for shifts
-                    //     emit_mov(self, R9, RCX);
-                    //     emit_call(self, syscall.function as usize);
-                    // } else {
-                    //     return Err(Error::new(ErrorKind::Other,
-                    //                    format!("[JIT] Error: unknown syscall function (id: {:#x})",
-                    //                            insn.imm as u32)));
-                    // };
+                ebpf::CALL_IMM   => {
+                    // Save registers for call before displacement
+                    emit_register_saving_before_call(self);
+
+                    // For JIT, syscalls MUST be registered at compile time. They can be
+                    // updated later, but not created after compiling (we need the address of the
+                    // syscall function in the JIT-compiled program).
+                    let func_ptr = match syscalls.get(&(insn.imm as u32)) {
+                        Some(Syscall::Function(func)) => *func as *const u8,
+                        Some(Syscall::Object(boxed)) => {
+                            let fat_ptr_ptr = unsafe { std::mem::transmute::<_, *const *const SyscallTraitObject>(&boxed) };
+                            let fat_ptr = unsafe { std::mem::transmute::<_, *const SyscallTraitObject>(*fat_ptr_ptr) };
+                            // Displace arguments to make room for RSI
+                            for i in (2..6).rev() {
+                                emit_mov(self, ARGUMENT_REGISTERS[i-1], ARGUMENT_REGISTERS[i]);
+                            }
+                            // RSI is the "&mut self" in the "call" method of the SyscallObject
+                            emit_load_imm(self, RSI, unsafe { (*fat_ptr).data } as i64);
+                            let vtable = unsafe { std::mem::transmute::<_, &SyscallObjectVtable>(&*(*fat_ptr).vtable) };
+                            vtable.call
+                        },
+                        None => {
+                            return Err(JITError::UnknownSyscall(insn.imm as u32, insn_ptr));
+                        }
+                    };
+
+                    // emit_mov(self, R9, R10); // TODO: Pass mem / regions
```

Comment:
Next steps?

Reply:
The memory translation in JIT is incompatible with the one from the interpreter anyway and needs to be redesigned. Thus, I left this part out and will address it in the next PR.